File: class/practice5.2/MegaData.cpp
#include "MegaData.h"
#include <algorithm>
#include <iostream>

constexpr size_t SMALL_ARRAY_SIZE = 1024;
constexpr size_t BIG_ARRAY_SIZE = 1024 * 1024;

MegaData::MegaData()
        : smallArray(SMALL_ARRAY_SIZE, 42.0f), bigArray(BIG_ARRAY_SIZE, 42.0) {}

void MegaData::reset() {
    std::fill(smallArray.begin(), smallArray.end(), 42.0f);
    std::fill(bigArray.begin(), bigArray.end(), 42.0);
}

MegaDataPool::MegaDataPool(size_t size) {
pool.reserve(size);
for (size_t i = 0; i < size; ++i) {
pool.push_back(std::make_unique<MegaData>());
available.push_back(pool.back().get());
}
}

MegaDataPool& MegaDataPool::getInstance(size_t size) {
static MegaDataPool instance(size);
return instance;
}

MegaData* MegaDataPool::acquire() {
    std::lock_guard<std::mutex> lock(poolMutex);
    if (available.empty()) {
        return nullptr;
    }
    MegaData* obj = available.back();
    available.pop_back();
    return obj;
}

void MegaDataPool::release(MegaData* obj) {
    std::lock_guard<std::mutex> lock(poolMutex);
    obj->reset();
    available.push_back(obj);
}

size_t MegaDataPool::size() const {
    return pool.size();
}

size_t MegaDataPool::usedSize() const {
    std::lock_guard<std::mutex> lock(poolMutex);
    return pool.size() - available.size();
}

MegaData* acquireObjectFromPool() {
    return MegaDataPool::getInstance(10).acquire();
}

void releaseObjectToPool(MegaData* obj) {
    MegaDataPool::getInstance(10).release(obj);
}

size_t getPoolUsedSize() {
    return MegaDataPool::getInstance(10).usedSize();
}

File: home/home5.1/src/MegaData.java
public class MegaData {
    private static final int SMALL_ARRAY_SIZE = 1024;
    private static final int BIG_ARRAY_SIZE = 1024 * 1024;

    public float[] smallArray;
    public double[] bigArray;

    public MegaData() {
        smallArray = new float[SMALL_ARRAY_SIZE];
        bigArray = new double[BIG_ARRAY_SIZE];
        reset();
    }

    public void reset() {
        for (int i = 0; i < SMALL_ARRAY_SIZE; i++) {
            smallArray[i] = 42.0f;
        }
        for (int i = 0; i < BIG_ARRAY_SIZE; i++) {
            bigArray[i] = 42.0;
        }
    }
}
